### PR TITLE
Let DirectMethodReceiver reports back to TRC

### PR DESCRIPTION
**Pull request by yophilav**
- Modify the Direct Method Receiver to report back to Test Result Coordinator (If applicable)
- DMRc is now responsible for generating report that is compliance with DirectMethod reporting type

### CODE REVIEW

File: test/modules/DirectMethodReceiver/src/DirectMethodReceiver.cs [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    using System;
+    using System.Net;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.Devices.Client;
+    using Microsoft.Azure.Devices.Edge.ModuleUtil;
+    using Microsoft.Azure.Devices.Edge.ModuleUtil.TestResults;
+    using Microsoft.Azure.Devices.Edge.Util;
+    using Microsoft.Extensions.Configuration;
+    using Microsoft.Extensions.Logging;
+    class DirectMethodReceiver : IDisposable
```
Comment (philipktlin) [inline test/modules/DirectMethodReceiver/src/DirectMethodReceiver.cs]: nit: line break.

File: test/modules/DirectMethodReceiver/src/DirectMethodReceiver.cs [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+            }
+
+            return new DirectMethodReceiver(
+                logger,
+                configuration,
+                moduleClient,
+                testResultReportingClient);
+        }
+
+        public void Dispose() => this.moduleClient?.Dispose();
```
Comment (philipktlin) [inline test/modules/DirectMethodReceiver/src/DirectMethodReceiver.cs]: Since moduleClient is passing in, clean up should be handled by caller.

File: test/modules/DirectMethodReceiver/src/Program.cs [hunk 3]
```diff
@@ -20,37 +18,41 @@ class Program
...
+                    .SetBasePath(Directory.GetCurrentDirectory())
+                    .AddJsonFile("config/appsettings.json", optional: true)
+                    .AddEnvironmentVariables()
+                    .Build();
+
+                directMethodClient = await DirectMethodReceiver.CreateAsync(
+                    Logger,
+                    configuration);
+
+                (CancellationTokenSource cts, ManualResetEventSlim completed, Option<object> handler) = ShutdownHandler.Init(TimeSpan.FromSeconds(5), Logger);
```
Comment (philipktlin) [inline test/modules/DirectMethodReceiver/src/Program.cs]: nit: Move this line to line 25.

File: test/modules/DirectMethodReceiver/src/DirectMethodReceiver.cs [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+        }
+
+        public static async Task<DirectMethodReceiver> CreateAsync(
+            ILogger logger,
+            IConfiguration configuration)
+        {
+            Preconditions.CheckNotNull(logger, nameof(logger));
+            Preconditions.CheckNotNull(configuration, nameof(configuration));
+
+            ModuleClient moduleClient = await ModuleUtil.CreateModuleClientAsync(
```
Comment (philipktlin) [inline test/modules/DirectMethodReceiver/src/DirectMethodReceiver.cs]: Move module client creation inside constructor

Comment (yophilav) [reply test/modules/DirectMethodReceiver/src/DirectMethodReceiver.cs]: Can a constructor be async?

File: test/modules/DirectMethodReceiver/src/DirectMethodReceiver.cs [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+                    SequenceNumber = this.directMethodCount.ToString(),
+                    Result = HttpStatusCode.OK.ToString()
+                };
+            }
+
+            await ModuleUtil.ReportTestResultAsync(this.testResultReportingClient, this.logger, testResult);
+            this.directMethodCount++;
+        }
+
+        public async Task StartAsync()
```
Comment (philipktlin) [inline test/modules/DirectMethodReceiver/src/DirectMethodReceiver.cs]: I think it is better to move it inside constructor.

File: test/modules/DirectMethodReceiver/src/Program.cs [hunk 6]
```diff
@@ -20,37 +18,41 @@ class Program
...
-            await moduleClient.SetMethodHandlerAsync("HelloWorldMethod", HelloWorldMethodAsync, null);
-            await cts.Token.WhenCanceled();
-
-            completed.Set();
-            handler.ForEach(h => GC.KeepAlive(h));
-            Logger.LogInformation("DirectMethodReceiver Main() finished.");
-            return 0;
-        }
+            DirectMethodReceiver directMethodClient = null;
+            try
```
Comment (philipktlin) [inline test/modules/DirectMethodReceiver/src/Program.cs]:
```suggestion

            try
```

File: test/modules/DirectMethodReceiver/src/DirectMethodReceiver.cs [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+                {
+                    TrackingId = Option.Maybe(this.configuration.GetValue<string>("trackingId"))
+                        .Expect(() => new ArgumentException("TrackingId is empty")),
+                    BatchId = this.batchId.ToString(),
+                    SequenceNumber = this.directMethodCount.ToString(),
+                    Result = HttpStatusCode.OK.ToString()
+                };
+            }
+
+            await ModuleUtil.ReportTestResultAsync(this.testResultReportingClient, this.logger, testResult);
```
Comment (philipktlin) [inline test/modules/DirectMethodReceiver/src/DirectMethodReceiver.cs]: this should move inside if block, right?

File: test/modules/DirectMethodReceiver/src/DirectMethodReceiver.cs (hunk 5 above)
Comment (yophilav) [reply test/modules/DirectMethodReceiver/src/DirectMethodReceiver.cs]: constructor cannot be async.

File: test/modules/DirectMethodReceiver/src/DirectMethodReceiver.cs [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+            this.directMethodCount++;
+        }
+
+        public async Task StartAsync()
+        {
+            await this.moduleClient.OpenAsync();
+            await this.moduleClient.SetMethodHandlerAsync("HelloWorldMethod", this.HelloWorldMethodAsync, null);
```
Comment (philipktlin) [inline test/modules/DirectMethodReceiver/src/DirectMethodReceiver.cs]: use synchronous version, https://docs.microsoft.com/en-us/dotnet/api/microsoft.azure.devices.client.deviceclient.setmethodhandler?view=azure-dotnet.

File: test/modules/DirectMethodReceiver/src/DirectMethodReceiver.cs (hunk 5 above)
Comment (philipktlin) [inline test/modules/DirectMethodReceiver/src/DirectMethodReceiver.cs]: nit: RegisterDirectMethodHandler / RegisterDirectMethodCallback

File: test/modules/DirectMethodReceiver/src/DirectMethodReceiver.cs [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+        public DirectMethodReceiver(
+            ILogger logger,
+            IConfiguration configuration)
+        {
+            Preconditions.CheckNotNull(logger, nameof(logger));
+            Preconditions.CheckNotNull(configuration, nameof(configuration));
+
+            TestResultReportingClient testResultReportingClient = null;
+            Option<Uri> testReportCoordinatorUrl = Option.Maybe(configuration.GetValue<Uri>("testResultCoordinatorUrl"));
+            if (testReportCoordinatorUrl.HasValue)
```
Comment (philipktlin) [inline test/modules/DirectMethodReceiver/src/DirectMethodReceiver.cs]: nit: use testReportCoordinatorUrl.ForEach and make testResultReportingClient as Option<TestResultReportingClient>

File: test/modules/DirectMethodReceiver/src/DirectMethodReceiver.cs [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+            this.logger.LogInformation("Received direct method call.");
+            // Send the report to Test Result Coordinator
+            await this.ReportTestResult();
+            return new MethodResponse((int)HttpStatusCode.OK);
+        }
+
+        public async Task ReportTestResult()
+        {
+            DirectMethodTestResult testResult = null;
+            if (this.testResultReportingClient != null)
```
Comment (philipktlin) [inline test/modules/DirectMethodReceiver/src/DirectMethodReceiver.cs]: nit: use testResultReportingClient.ForEach

File: test/modules/DirectMethodReceiver/src/DirectMethodReceiver.cs [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+    using Microsoft.Azure.Devices.Edge.ModuleUtil.TestResults;
+    using Microsoft.Azure.Devices.Edge.Util;
+    using Microsoft.Extensions.Configuration;
+    using Microsoft.Extensions.Logging;
+
+    class DirectMethodReceiver : IDisposable
+    {
+        Guid batchId;
+        IConfiguration configuration;
+        long directMethodCount = 1;
```
Comment (philipktlin) [inline test/modules/DirectMethodReceiver/src/DirectMethodReceiver.cs]: start from 0

File: test/modules/DirectMethodReceiver/src/DirectMethodReceiver.cs [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+                    TrackingId = Option.Maybe(this.configuration.GetValue<string>("trackingId"))
+                        .Expect(() => new ArgumentException("TrackingId is empty")),
+                    BatchId = this.batchId.ToString(),
+                    SequenceNumber = this.directMethodCount.ToString(),
+                    Result = HttpStatusCode.OK.ToString()
+                };
+                await ModuleUtil.ReportTestResultAsync(this.testResultReportingClient, this.logger, testResult);
+            }
+
+            this.directMethodCount++;
```
Comment (philipktlin) [inline test/modules/DirectMethodReceiver/src/DirectMethodReceiver.cs]: move this line into HelloWorldMethodAsync method

File: test/modules/DirectMethodReceiver/src/DirectMethodReceiver.cs [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+        }
+
+        public async Task Init()
```
Comment (philipktlin) [inline test/modules/DirectMethodReceiver/src/DirectMethodReceiver.cs]:
```suggestion
        public async Task InitAsync()
```

File: test/modules/DirectMethodReceiver/src/DirectMethodReceiver.cs [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+        {
+            this.moduleClient = await ModuleUtil.CreateModuleClientAsync(
+                this.configuration.GetValue("ClientTransportType", TransportType.Amqp_Tcp_Only),
+                ModuleUtil.DefaultTimeoutErrorDetectionStrategy,
+                ModuleUtil.DefaultTransientRetryStrategy,
+                this.logger);
+            Preconditions.CheckNotNull(this.moduleClient, nameof(this.moduleClient));
```
Comment (philipktlin) [inline test/modules/DirectMethodReceiver/src/DirectMethodReceiver.cs]: you don't need this line anymore.

File: test/modules/DirectMethodReceiver/src/Program.cs [hunk 15]
```diff
@@ -20,37 +18,42 @@ class Program
...
+
+                (CancellationTokenSource cts, ManualResetEventSlim completed, Option<object> handler) = ShutdownHandler.Init(TimeSpan.FromSeconds(5), Logger);
+
+                IConfiguration configuration = new ConfigurationBuilder()
+                    .SetBasePath(Directory.GetCurrentDirectory())
+                    .AddJsonFile("config/appsettings.json", optional: true)
+                    .AddEnvironmentVariables()
+                    .Build();
+
+                directMethodClient = new DirectMethodReceiver(
```
Comment (philipktlin) [inline test/modules/DirectMethodReceiver/src/Program.cs]: nit: make it into a single line

File: test/modules/ModuleLib/ModuleUtil.cs [hunk 16]
```diff
@@ -61,7 +61,7 @@ public static async Task ReportTestResultAsync(TestResultReportingClient apiClie
             try
             {
                 logger.LogInformation($"Sending test result: Source={testResult.Source}, Type={testResult.ResultType}, CreatedAt={testResult.CreatedAt}, Result={testResult.GetFormattedResult()}");
-                await apiClient.ReportResultAsync(testResult.ToTestOperationResultDto());
+                await apiClient?.ReportResultAsync(testResult?.ToTestOperationResultDto());
```
Comment (philipktlin) [inline test/modules/ModuleLib/ModuleUtil.cs]: I think it is no longer needed.

Comment (yophilav) [reply test/modules/ModuleLib/ModuleUtil.cs]: I doesn't do any harm to use the ?. I'm going to just leave it here.

File: test/modules/DirectMethodReceiver/src/DirectMethodReceiver.cs (hunk 12 above)
Comment (yophilav) [reply test/modules/DirectMethodReceiver/src/DirectMethodReceiver.cs]: good catch! It was there before then I move things around and that got left behind

File: test/modules/DirectMethodReceiver/src/DirectMethodReceiver.cs (hunk 11 above)
Comment (yophilav) [reply test/modules/DirectMethodReceiver/src/DirectMethodReceiver.cs]: The DM sender has always start at 1. I'm keeping the consistency here.

File: test/modules/DirectMethodReceiver/src/DirectMethodReceiver.cs (hunk 10 above)
Comment (yophilav) [reply test/modules/DirectMethodReceiver/src/DirectMethodReceiver.cs]: testResultReportingClient  is not a type of Option<>

File: edgelet/build/windows/util.ps1 [hunk 17]
```diff
@@ -10,7 +10,9 @@ function Test-RustUp
 
 function GetPrivateRustPath
 {
-    Join-Path -Path (Get-IotEdgeFolder) -ChildPath 'rust-windows-arm/rust-windows-arm/bin/'
+    $toolchain = Get-Content -Encoding UTF8 (Join-Path -Path (Get-EdgeletFolder) -ChildPath 'rust-toolchain')
```
Comment (philipktlin) [inline edgelet/build/windows/util.ps1]: this file's changes are not related to this PR.

File: test/modules/DirectMethodReceiver/src/DirectMethodReceiver.cs [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+        public DirectMethodReceiver(
+            ILogger logger,
+            IConfiguration configuration)
+        {
+            Preconditions.CheckNotNull(logger, nameof(logger));
+            Preconditions.CheckNotNull(configuration, nameof(configuration));
+
+            TestResultReportingClient testResultReportingClient = null;
+            Option<Uri> testReportCoordinatorUrl = Option.Maybe(configuration.GetValue<Uri>("testResultCoordinatorUrl"));
+            testReportCoordinatorUrl.ForEach(
```
Comment (philipktlin) [inline test/modules/DirectMethodReceiver/src/DirectMethodReceiver.cs]: nit: ForEach supports 2nd argument to provide none action.  You can assign this.testResultReportingClient directly inside ForEach block.

File: test/modules/DirectMethodReceiver/src/DirectMethodReceiver.cs [hunk 19]
```diff
@@ -0,0 +1,82 @@
...
+            this.logger.LogInformation("Received direct method call.");
+            // Send the report to Test Result Coordinator
+            await this.ReportTestResult();
+            this.directMethodCount++;
+            return new MethodResponse((int)HttpStatusCode.OK);
+        }
+
+        public async Task ReportTestResult()
+        {
+            DirectMethodTestResult testResult = null;
```
Comment (philipktlin) [inline test/modules/DirectMethodReceiver/src/DirectMethodReceiver.cs]: combine declaration and assignment.

File: test/modules/DirectMethodReceiver/src/DirectMethodReceiver.cs [hunk 20]
```diff
@@ -0,0 +1,82 @@
...
+            await this.testResultReportingClient.ForEachAsync(
+                async (TestResultReportingClient testResultReportingClient) =>
+                    {
+                        testResult = new DirectMethodTestResult(this.configuration.GetValue<string>("IOTEDGE_MODULEID") + ".receive", DateTime.UtcNow)
+                        {
+                            TrackingId = Option.Maybe(this.configuration.GetValue<string>("trackingId"))
```
Comment (philipktlin) [inline test/modules/DirectMethodReceiver/src/DirectMethodReceiver.cs]: define a class field for tracking id and assign it in constructor.

File: test/modules/ModuleLib/ModuleUtil.cs (hunk 16 above)
Comment (philipktlin) [reply test/modules/ModuleLib/ModuleUtil.cs]: why do you need to call apiClient if it is null?  and why do you want to report result if test result is null?  Please revert this.

File: edgelet/build/windows/util.ps1 (hunk 17 above)
Comment (yophilav) [reply edgelet/build/windows/util.ps1]: Not it doesn't but I need it to build successfully for testing.

File: test/modules/DirectMethodReceiver/src/DirectMethodReceiver.cs (hunk 20 above)
Comment (yophilav) [reply test/modules/DirectMethodReceiver/src/DirectMethodReceiver.cs]: I like that. 👍

File: test/modules/DirectMethodReceiver/src/Program.cs [hunk 21]
```diff
@@ -20,37 +18,40 @@ class Program
...
-            await moduleClient.OpenAsync();
-            await moduleClient.SetMethodHandlerAsync("HelloWorldMethod", HelloWorldMethodAsync, null);
-            await cts.Token.WhenCanceled();
-
-            completed.Set();
-            handler.ForEach(h => GC.KeepAlive(h));
-            Logger.LogInformation("DirectMethodReceiver Main() finished.");
-            return 0;
-        }
+            DirectMethodReceiver directMethodClient = null;
```
Comment (philipktlin) [inline test/modules/DirectMethodReceiver/src/Program.cs]: Sorry, missed this.
```suggestion
            DirectMethodReceiver directMethodReceiver = null;
```

File: test/modules/DirectMethodReceiver/src/DirectMethodReceiver.cs [hunk 22]
```diff
@@ -0,0 +1,83 @@
...
+
+            Option<Uri> testReportCoordinatorUrl = Option.Maybe(configuration.GetValue<Uri>("testResultCoordinatorUrl"));
+            testReportCoordinatorUrl.ForEach(
+                (Uri uri) => this.testResultReportingClient = Option.Some<TestResultReportingClient>(new TestResultReportingClient { BaseUrl = uri.AbsoluteUri }),
+                () => this.testResultReportingClient = Option.None<TestResultReportingClient>());
+
+            this.logger = logger;
+            this.configuration = configuration;
+            this.batchId = Guid.NewGuid();
+            this.trackingId = Option.Maybe(this.configuration.GetValue<string>("trackingId"));
```
Comment (philipktlin) [inline test/modules/DirectMethodReceiver/src/DirectMethodReceiver.cs]: You can validate if testReportCoordinatorUrl have value; therefore below usage, you can call OrDefault() instead of Expect.
